proof_generation/proof_generation_utils.js

```diff
@@ -0,0 +1,538 @@
+   *
+   * @returns {Object} An object containing the deployed contract address.
+   */
+  getContractAddresses() {
+    const addresses = {};
+    Object.keys(this.registeredContracts).map((key) => {
+      if (this.registeredContracts[key].address) {
+        addresses[key] = this.registeredContracts[key].address;
+      }
+    });
+    return addresses;
+  }
+
+  /**
+   * Resets proof data variable.
+   */
+  resetProofData() {
+    this.proofData = {};
+  }
+
+  /**
+   * Initialize proof data with default data.
+   *
+   * @returns {Promise<void>}
+   */
+  async initializeProofData() {
+    this.proofData.contracts = this.getContractAddresses(this.registeredContracts);
+    this.proofData.gateway = {};
+    this.proofData.gateway.constructor = await this.gateway.getConstructorParams();
+    this.proofData.co_gateway = {};
+    this.proofData.co_gateway.constructor = await this.coGateway.getConstructorParams();
+  }
+
+  /**
+   * Populates proof data after calling stake.
+   *
+   * @returns {Promise<{stakeProofData, stakeResult: void}>}
+   */
+  async populateStakeProofData() {
+    const generatedHashLock = TestLibUtils.generateHashLock();
+    this.stakeParams.hashLock = generatedHashLock.l;
+    this.stakeParams.unlockSecret = generatedHashLock.s;
+
+    const stakeResult = await this.gateway.stake(this.stakeParams);
+    const stakeProofData = await this.proofUtils.getOutboxProof(
+      this.gateway.address,
+      [stakeResult.returned_value.messageHash_],
+    );
+    this.proofData.gateway.stake = {};
+    this.proofData.gateway.stake.params = this.stakeParams;
+    this.proofData.gateway.stake.return_value = stakeResult;
+    this.proofData.gateway.stake.proof_data = stakeProofData;
+    return {
+      stakeProofData,
+      stakeResult,
+    };
+  }
+
+  /**
+   * Populates proof data after calling confirmStakingIntent.
+   *
+   * @param stakeProofData {object} Contains proof data after calling stake.
+   *
+   * @returns {Promise<{confirmStakeIntentResult}>}
+   */
+  async populateConfirmStakeIntentProofData(stakeProofData) {
+    const confirmStakeIntentParams = Object.assign({}, this.stakeParams);
+    confirmStakeIntentParams.blockHeight = stakeProofData.block_number;
+    confirmStakeIntentParams.rlpParentNodes = stakeProofData.storageProof[0].serializedProof;
+    confirmStakeIntentParams.facilitator = this.stakeParams.staker;
+    confirmStakeIntentParams.storageRoot = stakeProofData.storageHash;
+
+    // confirmStakeIntent also sets/anchors storage root for a block number.
+    const confirmStakeIntentResult = await this.coGateway.confirmStakeIntent(
+      confirmStakeIntentParams,
+    );
+    const confirmStakeIntentProofData = await this.proofUtils.getInboxProof(
+      this.coGateway.address,
+      [confirmStakeIntentResult.returned_value.messageHash_],
+    );
+    this.proofData.co_gateway.confirm_stake_intent = {};
+    this.proofData.co_gateway.confirm_stake_intent.params = confirmStakeIntentParams;
```

Comment:
Maybe it makes sense to extra object into another variable in favor of short statement and readablity. 🤔 Also at other places. 

```
    this.proofData.co_gateway.confirm_stake_intent = {};
    const confirmStakeIntent = this.proofData.co_gateway.confirm_stake_intent;
    confirmStakeIntent.params = confirmStakeIntentParams;
    confirmStakeIntent.return_value = confirmStakeIntentResult;
    confirmStakeIntent.proof_data = confirmStakeIntentProofData;

```

Reply:
proofData is structured in a way it will be written in JSON file. That's why the nesting of parameters and representation. 
If we do as suggested I think It will be confusing how data is being written/nested in this.proof_data variable which might not be readable in the first place  💭 💭 

Thoughts 🤔 ?

Reply:
Please ignore my above comment. Updated with below:
```
const confirmStakeIntent = {};
    confirmStakeIntent.params = confirmStakeIntentParams;
    confirmStakeIntent.return_value = confirmStakeIntentResult;
    confirmStakeIntent.proof_data = confirmStakeIntentProofData;
    this.proofData.co_gateway.confirm_stake_intent = confirmStakeIntent;
```